homeassistant/components/google_travel_time/config_flow.py

```diff
@@ -0,0 +1,152 @@
+
+        if CONF_ARRIVAL_TIME in self.config_entry.options:
+            default_time_type = ARRIVAL_TIME
+            default_time = self.config_entry.options[CONF_ARRIVAL_TIME]
+        else:
+            default_time_type = DEPARTURE_TIME
+            default_time = self.config_entry.options.get(CONF_ARRIVAL_TIME)
+
+        return self.async_show_form(
+            step_id="init",
+            data_schema=vol.Schema(
+                {
+                    vol.Optional(
+                        CONF_MODE, default=self.config_entry.options[CONF_MODE]
+                    ): vol.In(TRAVEL_MODE),
+                    vol.Optional(
+                        CONF_LANGUAGE,
+                        default=self.config_entry.options.get(CONF_LANGUAGE),
+                    ): vol.In(ALL_LANGUAGES),
+                    vol.Optional(
+                        CONF_AVOID, default=self.config_entry.options.get(CONF_AVOID)
+                    ): vol.In(AVOID),
+                    vol.Optional(
+                        CONF_UNITS, default=self.config_entry.options[CONF_UNITS]
+                    ): vol.In(UNITS),
+                    vol.Optional(CONF_TIME_TYPE, default=default_time_type): vol.In(
+                        TIME_TYPES
+                    ),
+                    vol.Optional(CONF_TIME, default=default_time): cv.string,
+                    vol.Optional(
+                        CONF_TRAFFIC_MODEL,
+                        default=self.config_entry.options.get(CONF_TRAFFIC_MODEL),
+                    ): vol.In(TRAVEL_MODEL),
+                    vol.Optional(
+                        CONF_TRANSIT_MODE,
+                        default=self.config_entry.options.get(CONF_TRANSIT_MODE),
+                    ): vol.In(TRANSPORT_TYPE),
+                    vol.Optional(
+                        CONF_TRANSIT_ROUTING_PREFERENCE,
+                        default=self.config_entry.options.get(
+                            CONF_TRANSIT_ROUTING_PREFERENCE
+                        ),
+                    ): vol.In(TRANSIT_PREFS),
+                }
+            ),
+        )
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Google Maps Travel Time."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(
+        config_entry: config_entries.ConfigEntry,
+    ) -> GoogleOptionsFlow:
+        """Get the options flow for this handler."""
+        return GoogleOptionsFlow(config_entry)
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+        if user_input is not None:
+            await self.async_set_unique_id(
+                slugify(
+                    f"{DOMAIN}_{user_input[CONF_ORIGIN]}_{user_input[CONF_DESTINATION]}"
+                )
+            )
+            self._abort_if_unique_id_configured()
+            return self.async_create_entry(
```

Comment:
We should validate the api key before creating the entry.

Reply:
That's new functionality. Does that have to be part of this PR or can we make it part of a follow up one?

Reply:
It's not really new since we validate the api key in the platform. We always must validate credentials in the config flow before creating the entry. That's standard.

Reply:
nvm I see we are doing a poor version of validation right now

Reply:
Hmm I think we should handle this in another PR. The reason is because this would require a significant refactor to fix. There's a lot of logic to resolve the locations from an entity or zone that would need to be pulled out of the sensor and made to be generic